File: core/core-enum/src/main/java/com/imstargg/core/enums/BattleType.java
package com.imstargg.core.enums;

import java.util.Arrays;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;

public enum BattleType {

    NOT_FOUND(null),

    RANKED("ranked"),
    SOLO_RANKED("soloRanked"),
    FRIENDLY("friendly"),
    CHALLENGE("challenge"),
    TOURNAMENT("tournament"),
    CHAMPIONSHIP_CHALLENGE("championshipChallenge"),
    ;

    private final String code;

    private static final Map<String, BattleType> ENUM_BY_CODE = Arrays.stream(BattleType.values())
            .filter(e -> e != NOT_FOUND)
            .collect(Collectors.toMap(BattleType::getCode, Function.identity()));

    private static final Set<BattleType> REGULAR_TYPES = Set.of(RANKED, SOLO_RANKED);

    public static BattleType find(String code) {
        return ENUM_BY_CODE.getOrDefault(code, NOT_FOUND);
    }

    public static Set<BattleType> regularTypes() {
        return REGULAR_TYPES;
    }

    BattleType(String code) {
        this.code = code;
    }

    public boolean isRegular() {
        return REGULAR_TYPES.contains(this);
    }

    public String getCode() {
        return code;
    }
}

File: imstargg-frontend/src/model/enums/BattleType.ts
export const BattleTypeValue = {
  NOT_FOUND: 'NOT_FOUND',

  RANKED: 'RANKED',
  SOLO_RANKED: 'SOLO_RANKED',
  FRIENDLY: 'FRIENDLY',
  CHALLENGE: 'CHALLENGE',
  TOURNAMENT: 'TOURNAMENT',
  CHAMPIONSHIP_CHALLENGE: 'CHAMPIONSHIP_CHALLENGE',
} as const;

export type BattleType = typeof BattleTypeValue[keyof typeof BattleTypeValue];

export const BattleTypeValues = Object.keys(BattleTypeValue) as BattleType[];

export const StatisticsBattleTypeValue = {
  ALL: 'ALL',
  RANKED: BattleTypeValue.RANKED,
  SOLO_RANKED: BattleTypeValue.SOLO_RANKED,
} as const;

export type StatisticsBattleType = typeof StatisticsBattleTypeValue[keyof typeof StatisticsBattleTypeValue];

export const StatisticsBattleTypeValues = Object.keys(StatisticsBattleTypeValue) as (keyof typeof StatisticsBattleTypeValue)[];